### ISSUE TITLE
`[allow(clippy::disallowed_macros)]` fails for proc-macro attributes

### ISSUE THREAD

**Issue by bnjbvr** [rust-lang/rust-clippy#13521]
### Summary

If I have a proc macro attribute, that's part of the `disallowed-macros` list, and I add an `#[allow(clippy::disallowed_macros)]` on top of usage of the macro, Clippy still detects it as usage of a disallowed macro.

### Lint Name

disallowed-macros

### Reproducer

As a simpler repro, I've used [this repo](https://github.com/nazmulidris/rust-scratch/) with a new `clippy.toml` file in `/macros` that contained:

```
disallowed-macros = [
    "my_proc_macros_lib::attrib_macro_logger_1"
]
```

I tried this code in `/macros/src/main.rs`:

```rust
#[allow(clippy::disallowed_macros)]
#[my_proc_macros_lib::attrib_macro_logger_1(key = "value")]
fn jeej() {}
```

I saw this happen:

```
warning: use of a disallowed macro `my_proc_macros_lib::attrib_macro_logger_1`
  --> src/main.rs:23:1
   |
23 | #[my_proc_macros_lib::attrib_macro_logger_1(key = "value")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_macros
   = note: `#[warn(clippy::disallowed_macros)]` on by default
```

I expected to see this happen:

(no clippy complaint about this particular use)

### Version

rustc 1.83.0-nightly (3ae715c8c 2024-10-07)
binary: rustc
commit-hash: 3ae715c8c63f9aeac47cbf7d8d9dadb3fa32c638
commit-date: 2024-10-07
host: x86_64-unknown-linux-gnu
release: 1.83.0-nightly
LLVM version: 19.1.1

**Comment by zihan0822** [comment]
Is there any chance that the attribute attached to `proc_macro` is silently ignored during expansion? Currently, when you add a `#[allow(clippy::disallowed_macro)]` attribute to a declarative macro invocation, rustc will throw :
```
the built-in attribute `allow` will be ignored, since it's applied to the macro invocation
```

You can probably use inner attribute `#![allow(clippy::disallowed_macro)]` to suppress the warning at mod level.

**Comment by 2asoft** [comment]
> Is there any chance that the attribute attached to proc_macro is silently ignored during expansion?

The proc macro is given all the attrs on the item it is on, including the diagnostic attrs.  In the case where the diagnostic attrs are swallowed, they are nonexistent by the time the lint level calculation happens.  The lint still runs, still finds the proc macro that was the origin of the expansion (because the lint can backtrace), so the lint can still fire.  But because there is no diagnostic attribute, lint level calculation defers to the outer level (the mod, as in your example).

I see a couple solutions:
1. Early in AST stage, before macro expansion, detect that the diagnostic attribute is "on" an attr macro, and store this detail in a new session store.
During HIR lowering (I think this would be the right time), merge attributes from this store with the other attributes that were in the AST.
Let lint level calculation proceed as normal.
2. Pre-macro-expansion, collect diagnostic attributes.  Proceed with expansion as normal.  Post-expansion, prepend the collected attributes to all items in the expanded fragment.

I believe the second solution is inferior, but it was much simpler to implement.
Anything worth doing is worth doing badly, so here's the badly done [fix for this](https://github.com/rust-lang/rust/pull/145914).

Can someone who knows (someone who knows) the area please take a look?

**Comment by Alexendoo** [comment]
For cases where you control the proc macro you can ensure it emits the attributes passed into it onto each item it produces, e.g. here https://github.com/nazmulidris/rust-scratch/blob/a3a2638ebf6b98207da6671939a1c6c4751f5471/macros/my_proc_macros_lib/src/logger.rs#L60-L66 in the initial example

If something is going to change here it would probably be on the `rustc` side of things since this also applies to e.g. `#[allow(unused)]`

**Comment by CommanderStorm** [comment]
@saberoueslati
you have self-assiged to a LOT of issues lately. Why don't you slow down a bit? Better 2-3 very high quality ones than 20 rushed ones 😉

**Comment by saberoueslati** [comment]
@CommanderStorm obviously your advice makes a lot of sense, I'm not rushing anything, most of the issues are pretty old and with no advances for months or even years so it's just a way for me to not lose sight of them that's all and I have extra time on hand in the coming weeks to work them slowly but surely.
